alibi_detect/utils/loading.py

```diff
@@ -0,0 +1,625 @@
+        raise NotImplementedError('Loading of pytorch optimizers not currently supported')
+    return optimizer
+
+
+def _get_nested_value(dic: dict, keys: list) -> Any:
+    """
+    Get a value from a nested dictionary.
+
+    Parameters
+    ----------
+    dic
+        The dictionary.
+    keys
+        List of keys to "walk" to nested value.
+        For example, to extract the value `dic['key1']['key2']['key3']`, set `keys = ['key1', 'key2', 'key3']`.
+
+    Returns
+    -------
+    The nested value specified by `keys`.
+    """
+    for key in keys:
+        try:
+            dic = dic[key]
+        except (TypeError, KeyError):
+            return None
+    return dic
+
+
+def _set_nested_value(dic: dict, keys: list, value: Any):
+    """
+    Set a value in a nested dictionary.
+
+    Parameters
+    ----------
+    dic
+        The dictionary.
+    keys
+        List of keys to "walk" to nested value.
+        For example, to set the value `dic['key1']['key2']['key3']`, set `keys = ['key1', 'key2', 'key3']`.
+    value
+        The value to set.
+    """
+    for key in keys[:-1]:
+        dic = dic.setdefault(key, {})
+    dic[keys[-1]] = value
+
+
+def read_config(filepath: Union[os.PathLike, str]) -> dict:
+    """
+    This function reads a detector toml config file and returns a dict specifying the detector.
+
+    Parameters
+    ----------
+    filepath
+        The filepath to the config.toml file.
+
+    Returns
+    -------
+        Parsed toml dictionary.
+    """
+    filepath = Path(filepath)
+
+    cfg = dict(toml.load(filepath))  # toml.load types return as MutableMapping, force to dict
+    logger.info('Loaded config file from %s', str(filepath))
+
+    # Convert keys in categories_per_feature back to str. This is undesirable, but currently necessary in order
+    # to use toml library, as error with toml dumping dict with int keys.
+    if 'categories_per_feature' in cfg:
+        new = {}
+        for key in cfg['categories_per_feature']:
+            new[int(key)] = cfg['categories_per_feature'][key]
+        cfg['categories_per_feature'] = new
+
+    # This is necessary as no None/null in toml spec., and missing values are set to defaults set in pydantic models.
+    # But we sometimes to do explicitly spec as None.
+    cfg = _replace(cfg, "None", None)
+
+    return cfg
+
+
+def resolve_cfg(cfg: dict, config_dir: Optional[Path]) -> dict:
+    """
+    Resolves artefacts in a config dict. For example x_ref='x_ref.npy' is resolved by loading the np.ndarray from
+    the .npy file. For a list of fields that are resolved, see
+    https://docs.seldon.io/projects/alibi-detect/en/latest/overview/config_file.html.
+
+    Parameters
+    ----------
+    cfg
+        The unresolved config dict.
+    config_dir
+        Filepath to directory the `config.toml` is located in. Only required if different from the
+        runtime directory, and artefacts are specified with filepaths relative to the config.toml file.
```

Comment:
What would be an example of using this functionality? Is this function intended for public usage as well?

Reply:
Do you mean using `resolve_config` in general? Or the `config_dir` kwarg?

To answer the first, I've left `read_config`, `resolve_config` and `validate_config` public so that advanced users can peform the `load_detector` workflow manually for troubleshooting i.e. `cfg = read_config()` -> `cfg = validate_config(cfg)` -> *custom fetching* -> `cfg = resolve_config(cfg)` -> `cfg = validate_config(cfg, resolved=True)`.

The above could also be useful if *custom fetching* of artefacts is required. However, the missing piece of the puzzle here is that `load_detector` currently only accepts a file path to a ` config.toml` (I originally had it accepting resolved `cfg` dicts too, so it could complete the above workflow. This is what I hand in mind as the full workflow if `config.toml` files contained refs to remote artefacts). We could make `resolve_config` etc private for now and come back to this at a later date if needed? Or we could just leave as is... 

Wrt to `config_dir`, this is used by default internally, since the we pass `load_detector` a `filepath` containing the `config.toml`, which would usually be different to the runtime directory i.e. `filepath='my_detector'` instead of `filepath='.'`.